### PR TITLE
TH2-2216. Added: check for message date when added to message batch, check for book id for sub events, checks for components of StoredMessageId and StoredTestEventId.

### PR DESCRIPTION
**Pull request by MrJonniK**

### CODE REVIEW

File: cradle-core/src/main/java/com/exactpro/cradle/testevents/TestEventSingleToStore.java [hunk 1]
```diff
@@ -70,8 +71,23 @@ public void setSuccess(boolean success)
...
+	{
+		if (ids == null)
+			return;
+		BookId eventBookId = getId().getBookId();
+		for (StoredMessageId id : ids)
+		{
+			BookId messageBookId = id.getBookId();
+			if (!eventBookId.equals(messageBookId))
+				throw new CradleStorageException("Book of message is '" +
+						messageBookId + "' that different from the current event book id '" + eventBookId + "'");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/testevents/TestEventSingleToStore.java:90]: "Book of message is (" + messageBookId + ") differs from the event book (" + eventBookId + ")"

File: cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java [hunk 2]
```diff
@@ -46,15 +47,19 @@ public static void validateMessage(CradleMessage message) throws CradleStorageEx
 	{
 		if (message.getId() == null)
 			throw new CradleStorageException("Messages ID cannot be null");
-			
-		if (message.getBookId() == null)
+		if (message.getBookId() == null || StringUtils.isEmpty(message.getBookId().toString()))
 			throw new CradleStorageException("Message must have a book");
-		if (StringUtils.isEmpty(message.getSessionAlias()))
+		if (StringUtils.isEmpty(message.getSessionAlias()) )
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java:52]: Extra space added?

File: cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java [hunk 3]
```diff
@@ -46,15 +47,19 @@ public static void validateMessage(CradleMessage message) throws CradleStorageEx
...
 			throw new CradleStorageException("Message must have a session alias");
 		if (message.getDirection() == null)
 			throw new CradleStorageException("Message must have a direction");
 		if (message.getTimestamp() == null)
 			throw new CradleStorageException("Message must have a timestamp");
+		Instant now = Instant.now();
+		if (message.getTimestamp().isAfter(now))
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java:59]: This check may lead to problem when validation is done not before adding a message. Is it possible?

File: cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java [hunk 4]
```diff
@@ -46,15 +47,19 @@ public static void validateMessage(CradleMessage message) throws CradleStorageEx
...
+			throw new CradleStorageException(
+					"Message timestamp '" + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
+							"' It can not be greater than current '" + TimeUtils.toLocalTimestamp(now) + "'");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java:62]: "can not" > "cannot"

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java (hunk 2 above)
Comment (MrJonniK) [reply cradle-core/src/main/java/com/exactpro/cradle/utils/MessageUtils.java:52]: This is my mistake

File: cradle-core/src/test/java/com/exactpro/cradle/messages/MessageBatchToStoreTest.java [hunk 5]
```diff
@@ -80,7 +83,11 @@ public void prepare()
 		return new Object[][]
 				{
 					{builder},                                                                                       //Empty message
+					{builder.bookId(null)},																			 //Null book
+					{builder.bookId(new BookId(""))},														 //Empty book
 					{builder.bookId(book)},                                                                          //Only book is set
+					{builder.bookId(book).sessionAlias(null).direction(null).timestamp(null)},               		 //Null session are set
+					{builder.bookId(book).sessionAlias("").direction(null).timestamp(null)},               		 	 //Empty session are set
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/test/java/com/exactpro/cradle/messages/MessageBatchToStoreTest.java:90]: "are set" > "is set".

File: cradle-core/src/main/java/com/exactpro/cradle/messages/MessageBatchToStore.java [hunk 6]
```diff
@@ -93,8 +96,14 @@ public StoredMessage addMessage(MessageToStore message) throws CradleStorageExce
...
 		
-		MessageUtils.validateMessage(message);  //Checking if book, session alias, direction, timestamp and content are set
-		
+		// Checking that the timestamp of a message is not from the future
+		// Other checks have already been done when the MessageToStore was created
+		Instant now = Instant.now();
+		if (message.getTimestamp().isAfter(now))
+			throw new CradleStorageException(
+					"Message timestamp '" + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
+							"' It cannot be greater than current '" + TimeUtils.toLocalTimestamp(now) + "'");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/messages/MessageBatchToStore.java:105]: "Message timestamp is " + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
		"' It cannot be greater than current timestamp (" + TimeUtils.toLocalTimestamp(now) + ")");

File: cradle-core/src/main/java/com/exactpro/cradle/testevents/TestEventSingleToStore.java [hunk 7]
```diff
@@ -70,8 +71,23 @@ public void setSuccess(boolean success)
...
+	{
+		if (ids == null)
+			return;
+		BookId eventBookId = getId().getBookId();
+		for (StoredMessageId id : ids)
+		{
+			BookId messageBookId = id.getBookId();
+			if (!eventBookId.equals(messageBookId))
+				throw new CradleStorageException("Book of message is (" +
+						messageBookId + ") differs from the event book (" + eventBookId + ")");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/testevents/TestEventSingleToStore.java:90]: throw new CradleStorageException("Book of message (" +
						messageBookId + ") differs from the event book (" + eventBookId + ")");

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/messages/MessageBatchToStore.java [hunk 8]
```diff
@@ -101,8 +101,8 @@ public StoredMessage addMessage(MessageToStore message) throws CradleStorageExce
 		Instant now = Instant.now();
 		if (message.getTimestamp().isAfter(now))
 			throw new CradleStorageException(
-					"Message timestamp '" + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
-							"' It cannot be greater than current '" + TimeUtils.toLocalTimestamp(now) + "'");
+					"Message timestamp is (" + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
+							") It cannot be greater than current timestamp (" + TimeUtils.toLocalTimestamp(now) + ")");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/messages/MessageBatchToStore.java:105]: "Message timestamp (" + TimeUtils.toLocalTimestamp(message.getTimestamp()) +
							") is greater than current timestamp (" + TimeUtils.toLocalTimestamp(now) + ")");

File: cradle-core/src/main/java/com/exactpro/cradle/utils/TestEventUtils.java [hunk 9]
```diff
@@ -80,23 +80,36 @@ public static void validateTestEvent(TestEvent event) throws CradleStorageExcept
...
 			throw new CradleStorageException("Test event must have a start timestamp");
-		
-		if (event.getEndTimestamp() != null && event.getEndTimestamp().isBefore(event.getStartTimestamp()))
-			throw new CradleStorageException("Test event cannot end sooner than it started");
-		
+		Instant now = Instant.now();
+		if (event.getStartTimestamp().isAfter(now))
+			throw new CradleStorageException(
+					"Event start timestamp '" + TimeUtils.toLocalTimestamp(event.getStartTimestamp()) +
+							"' It can not be greater than current '" + TimeUtils.toLocalTimestamp(now) + "'");
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/utils/TestEventUtils.java:93]: "Event start timestamp (" + TimeUtils.toLocalTimestamp(event.getStartTimestamp()) +
							") is greater than current timestamp (" + TimeUtils.toLocalTimestamp(now) + ")");

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

**Pull request merged**